docs/contributing/BuildFromSource.md

```diff
@@ -0,0 +1,147 @@
+* Install yarn globally (`npm install -g yarn`)
+
+### macOS/Linux
+
+Building ASP.NET Core on macOS or Linux requires:
+
+* If using macOS, you need macOS Sierra or newer.
+* If using Linux, you need a machine with all .NET Core Linux prerequisites: <https://docs.microsoft.com/en-us/dotnet/core/linux-prerequisites>
+* At least 10 GB of disk space and a good internet connection (our build scripts download a lot of tools and dependencies)
+* curl <https://curl.haxx.se> or Wget <https://www.gnu.org/software/wget>
+* Git <https://git-scm.org>
+* NodeJS. LTS version of 10.14.2 or newer <https://nodejs.org>
+
+**NOTE** some ISPs have been know to use web filtering software that has caused issues with git repository cloning, if you experience issues cloning this repo please review <https://help.github.com/en/github/authenticating-to-github/using-ssh-over-the-https-port>
+
+## Building in Visual Studio
+
+Before opening the `src/Razor/Razor.sln` file in Visual Studio or VS Code, you need to perform the following actions.
+
+1. Executing the following on command-line:
+
+   ```ps1
+   .\restore.cmd
+   ```
+
+   This will download the required tools and build the entire repository once.
+
+   > :bulb: Pro tip: you will also want to run this command after pulling large sets of changes. On the master
+   > branch, we regularly update the versions of .NET Core SDK required to build the repo.
+   > You will need to restart Visual Studio every time we update the .NET Core SDK.
+   > To allow executing the setup script, you may need to update the execution policy on your machine.
+   You can do so by running the `Set-ExecutionPolicy -ExecutionPolicy RemoteSigned -Scope CurrentUser` command
+   in PowerShell. For more information on execution policies, you can read the [execution policy docs](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.security/set-executionpolicy).
+
+2. Use the `startvs.cmd src/Razor/Razor.sln` script to open Visual Studio with the Razor solution. This script first sets the required
+environment variables.
+
+### Common error: Unable to locate the .NET Core SDK
+
+Executing `.\restore.cmd` or `.\build.cmd` may produce these errors:
+
+> error : Unable to locate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the installed version.
+> error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
+
+In most cases, this is because the option _Use previews of the .NET Core SDK_ in VS2019 is not checked. Start Visual Studio, go to _Tools > Options_ and check _Use previews of the .NET Core SDK_ under _Environment > Preview Features_.
+
+
+## Building with Visual Studio Code
+
+Using Visual Studio Code with this repo requires setting environment variables on command line first.
+Use these command to launch VS Code with the right settings.
+
+On Windows (requires PowerShell):
+
+```ps1
+.\activate.ps1
```

Comment:
Shouldn't this be `. .\activate.ps1`?

Reply:
For `aspnetcore` that seems to be required. For `aspnetcore-tooling` I always just use `./activate.ps1`

Reply:
I've been typing two extra characters like a chump this whole time! FYI below in the `dotnet` from the command line section it uses `. .\activate.ps1`. But if they both work it's kind of whatever.

Reply:
> FYI below in the dotnet from the command line section it uses . .\activate.ps1. 

Ah good catch, I'll address this in some upcoming PR. 😄

Reply:
Actually scratch that, I think you may be on to something. It should be `. .\activate.ps1` for scoping:
https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_scripts#script-scope-and-dot-sourcing

and our script should be enforcing this (like it does in `dotnet/aspnetcore`). We have an outdated version of `activate.ps1` in comparison to aspnetcore. Updating...